File: src/系统/09．表现系统/02．对话框系统/14．任务物品发放.lua
local ____lualib = require("lualib_bundle")
local __TS__StringSplit = ____lualib.__TS__StringSplit
local __TS__StringTrim = ____lualib.__TS__StringTrim
local __TS__StringSubstring = ____lualib.__TS__StringSubstring
local __TS__ParseInt = ____lualib.__TS__ParseInt
local ____exports = {}
---
-- @noSelfInFile
local jass = require("jass.common")
local ____require_result_0 = require("lib.扩展函数.物品相关函数.创建物品函数")
local _____521B_5EFA_7269_54C1_5E76_6CE8_518C_6392_6CC4_76D1_542C = ____require_result_0["创建物品并注册排泄监听"]
local _____7ED9_4E88_5355_4F4D_7269_54C1 = ____require_result_0["给予单位物品"]
local ____require_result_1 = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版")
local stringToFourCCSafe = ____require_result_1.stringToFourCCSafe
local ____require_result_2 = require("系统.09．表现系统.06．广播提示消息.index")
local _____5E7F_64AD_5355_4F4D_63D0_793A = ____require_result_2["广播单位提示"]
local GetUnitX = jass.GetUnitX
local GetUnitY = jass.GetUnitY
____exports["发放任务物品"] = function(unit, itemConfig)
    if unit == nil or unit == 0 or not itemConfig or itemConfig == "" then
        return 0
    end
    local _____53D1_653E_6570_91CF = 0
    local _____6709_6389_843D = false
    local _____914D_7F6E_5217_8868 = __TS__StringSplit(itemConfig, "|")
    do
        local i = 0
        while i < #_____914D_7F6E_5217_8868 do
            do
                local _____914D_7F6E = __TS__StringTrim(_____914D_7F6E_5217_8868[i + 1])
                if _____914D_7F6E == "" then
                    goto __continue5
                end
                local _____6570_91CF_5206_9694_4F4D_7F6E = (string.find(_____914D_7F6E, "*", nil, true) or 0) - 1
                local _____7269_54C1_4EE3_7801 = __TS__StringTrim(_____6570_91CF_5206_9694_4F4D_7F6E >= 0 and __TS__StringSubstring(_____914D_7F6E, 0, _____6570_91CF_5206_9694_4F4D_7F6E) or _____914D_7F6E)
                local _____6570_91CF = _____6570_91CF_5206_9694_4F4D_7F6E >= 0 and (__TS__ParseInt(
                    __TS__StringSubstring(_____914D_7F6E, _____6570_91CF_5206_9694_4F4D_7F6E + 1),
                    10
                ) or 0) or 1
                if _____6570_91CF < 1 then
                    _____6570_91CF = 1
                end
                local _____7269_54C1_7C7B_578BID = stringToFourCCSafe(_____7269_54C1_4EE3_7801)
                if _____7269_54C1_7C7B_578BID == 0 then
                    goto __continue5
                end
                do
                    local j = 0
                    while j < _____6570_91CF do
                        do
                            local item = _____521B_5EFA_7269_54C1_5E76_6CE8_518C_6392_6CC4_76D1_542C(
                                _____7269_54C1_7C7B_578BID,
                                GetUnitX(unit),
                                GetUnitY(unit)
                            )
                            if item == nil or item == 0 then
                                goto __continue10
                            end
                            if _____7ED9_4E88_5355_4F4D_7269_54C1(unit, item) then
                                _____53D1_653E_6570_91CF = _____53D1_653E_6570_91CF + 1
                            else
                                _____53D1_653E_6570_91CF = _____53D1_653E_6570_91CF + 1
                                _____6709_6389_843D = true
                            end
                        end
                        ::__continue10::
                        j = j + 1
                    end
                end
            end
            ::__continue5::
            i = i + 1
        end
    end
    if _____6709_6389_843D then
        _____5E7F_64AD_5355_4F4D_63D0_793A(unit, "物品栏已满，部分奖励物品掉落在脚下，请注意拾取。", 4000)
    end
    return _____53D1_653E_6570_91CF
end
return ____exports

File: TS/系统/09．表现系统/02．对话框系统/14．任务物品发放.ts
/** @noSelfInFile */

const jass = require("jass.common") as any;
const { 创建物品并注册排泄监听, 给予单位物品 } = require("lib.扩展函数.物品相关函数.创建物品函数") as {
  创建物品并注册排泄监听: (this: void, itemId: number, x: number, y: number) => any;
  给予单位物品: (this: void, unit: any, item: any) => boolean;
};
const { stringToFourCCSafe } = require("lib.扩展函数.封装函数.01．通用工具.01．FourCC转换安全版") as {
  stringToFourCCSafe: (this: void, rawId: string | undefined | null) => number;
};
const { 广播单位提示 } = require("系统.09．表现系统.06．广播提示消息.index") as {
  广播单位提示: (this: void, 来源单位: any, 文本: string, 持续时间?: number) => void;
};

const GetUnitX = jass.GetUnitX as (this: void, unit: any) => number;
const GetUnitY = jass.GetUnitY as (this: void, unit: any) => number;

export function 发放任务物品(this: void, unit: any, itemConfig: string | undefined): number {
  if (unit == null || unit === 0 || !itemConfig || itemConfig === "") return 0;
  let 发放数量 = 0;
  let 有掉落 = false;
  const 配置列表 = itemConfig.split("|");
  for (let i = 0; i < 配置列表.length; i++) {
    const 配置 = 配置列表[i].trim();
    if (配置 === "") continue;
    const 数量分隔位置 = 配置.indexOf("*");
    const 物品代码 = (数量分隔位置 >= 0 ? 配置.substring(0, 数量分隔位置) : 配置).trim();
    let 数量 = 数量分隔位置 >= 0 ? parseInt(配置.substring(数量分隔位置 + 1), 10) || 0 : 1;
    if (数量 < 1) 数量 = 1;
    const 物品类型ID = stringToFourCCSafe(物品代码);
    if (物品类型ID === 0) continue;
    for (let j = 0; j < 数量; j++) {
      const item = 创建物品并注册排泄监听(物品类型ID, GetUnitX(unit), GetUnitY(unit));
      if (item == null || item === 0) continue;
      if (给予单位物品(unit, item)) {
        发放数量 += 1;
      } else {
        // 物品栏已满：物品保留在英雄脚下，不销毁、不判失败。
        发放数量 += 1;
        有掉落 = true;
      }
    }
  }
  if (有掉落) {
    广播单位提示(unit, "物品栏已满，部分奖励物品掉落在脚下，请注意拾取。", 4000);
  }
  return 发放数量;
}

export {};
